pkg/terror/terror.go

```diff
@@ -0,0 +1,331 @@
+	ClassBinlogOp
+	ClassCheckpoint
+	ClassTaskCheck
+	ClassRelayAPI
+	ClassRelayUnit
+	ClassDumpUnit
+	ClassLoadUnit
+	ClassSyncUnit
+	ClassDMMaster
+	ClassDMWorker
+	ClassDMTracer
+)
+
+var errClass2Str = map[ErrClass]string{
+	ClassDatabase:   "database",
+	ClassFunctional: "functional",
+	ClassConfig:     "config",
+	ClassBinlogOp:   "binlog-op",
+	ClassCheckpoint: "checkpoint",
+	ClassTaskCheck:  "task-check",
+	ClassRelayAPI:   "relay-api",
+	ClassRelayUnit:  "relay-unit",
+	ClassDumpUnit:   "dump-unit",
+	ClassLoadUnit:   "load-unit",
+	ClassSyncUnit:   "sync-unit",
+	ClassDMMaster:   "dm-master",
+	ClassDMWorker:   "dm-worker",
+	ClassDMTracer:   "dm-tracer",
+}
+
+// String implements fmt.Stringer interface
+func (ec ErrClass) String() string {
+	if s, ok := errClass2Str[ec]; ok {
+		return s
+	}
+	return fmt.Sprintf("unknown error class: %d", ec)
+}
+
+// ErrScope represents the error occurs environment, such as upstream DB error,
+// downstream DB error, DM internal error etc.
+type ErrScope int
+
+// Error scopes
+const (
+	ScopeNotSet ErrScope = iota
+	ScopeUpstream
+	ScopeDownstream
+	ScopeInternal
+)
+
+var errScope2Str = map[ErrScope]string{
+	ScopeNotSet:     "not-set",
+	ScopeUpstream:   "upstream",
+	ScopeDownstream: "downstream",
+	ScopeInternal:   "internal",
+}
+
+// String implements fmt.Stringer interface
+func (es ErrScope) String() string {
+	if s, ok := errScope2Str[es]; ok {
+		return s
+	}
+	return fmt.Sprintf("unknown error scope: %d", es)
+}
+
+// ErrLevel represents the emergency level of a specific error type
+type ErrLevel int
+
+// Error levels
+const (
+	LevelLow ErrLevel = iota + 1
+	LevelMedium
+	LevelHigh
+)
+
+var errLevel2Str = map[ErrLevel]string{
+	LevelLow:    "low",
+	LevelMedium: "medium",
+	LevelHigh:   "high",
+}
+
+// String implements fmt.Stringer interface
+func (el ErrLevel) String() string {
+	if s, ok := errLevel2Str[el]; ok {
+		return s
+	}
+	return fmt.Sprintf("unknown error level: %d", el)
+}
+
+// Error implements error interface and add more useful fields
+type Error struct {
+	code     ErrCode
+	class    ErrClass
+	scope    ErrScope
+	level    ErrLevel
+	message  string
+	args     []interface{}
+	rawCause error
+	stack    errors.StackTracer
+}
+
+// New creates a new *Error instance
+func New(code ErrCode, class ErrClass, scope ErrScope, level ErrLevel, message string) *Error {
+	return &Error{
+		code:    code,
+		class:   class,
+		scope:   scope,
+		level:   level,
+		message: message,
+	}
+}
+
+// Code returns ErrCode
+func (e *Error) Code() ErrCode {
+	return e.code
+}
+
+// Class returns ErrClass
+func (e *Error) Class() ErrClass {
+	return e.class
+}
+
+// Scope returns ErrScope
+func (e *Error) Scope() ErrScope {
+	return e.scope
+}
+
+// Level returns ErrLevel
+func (e *Error) Level() ErrLevel {
+	return e.level
+}
+
+// Error implements error interface.
+func (e *Error) Error() string {
+	return fmt.Sprintf("[%d:%s:%s:%s] %s", e.code, e.class, e.scope, e.level, e.getMsg())
```

Comment:
may be it is difficult to search in log file, how about change to format like [ErrCode=%d][Class=%s]...